File: src/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1: return False
        stack = []
        mapping = { '(': ')', '{': '}', '[': ']' }

        for c in s:
            if c in mapping:
                stack.append(c)
            else:
                if len(stack) == 0: return False
                p = stack.pop()
                if mapping[p] != c:
                    return False

        return len(stack) == 0

File: src/0020-valid-parentheses/solution.cpp
class Solution {
public:
    bool isValid(string _s) {
        stack<char> s;
        for(char c: _s) {
            if (c == '(' || c == '{' || c == '[') 
                s.push(c);
            else {
                if (s.empty()) return false;

                char d = s.top(); s.pop();
                string e = string(1, d) + string(1, c);
                
                if (e == "()") continue;
                else if (e == "{}") continue;
                else if (e == "[]") continue;
                else return false;
            }
        }
        return s.empty();
    }
};
